nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,535 @@
+    // the will cause an error if it will not set
+    nav2_util::declare_parameter_if_not_declared(
+      node, "polygons", rclcpp::PARAMETER_STRING_ARRAY);
+    nav2_util::declare_parameter_if_not_declared(
+      node, "observation_sources", rclcpp::PARAMETER_STRING_ARRAY);
+
+    // Working with polygons parameters
+    polygon_names_ = get_parameter("polygons").as_string_array();
+    for (std::string polygon_name : polygon_names_) {
+      // Leave it not initialized: the will cause an error if it will not set
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name + ".type", rclcpp::PARAMETER_STRING);
+      const std::string polygon_type = get_parameter(polygon_name + ".type").as_string();
+
+      if (polygon_type == "polygon") {
+        std::shared_ptr<Polygon> p = std::make_shared<Polygon>(
+          node, polygon_name, simulation_time_step);
+
+        if (!p->getParameters()) {
+          return false;
+        }
+
+        polygons_.push_back(p);
+      } else {  // polygon_type == "circle"
+        std::shared_ptr<Circle> c = std::make_shared<Circle>(
+          node, polygon_name, simulation_time_step);
+
+        if (!c->getParameters()) {
+          return false;
+        }
+
+        polygons_.push_back(c);
+      }
+    }
+
+    // Working with data sources parameters
+    source_names_ = get_parameter("observation_sources").as_string_array();
+    for (std::string source_name : source_names_) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, source_name + ".type",
+        rclcpp::ParameterValue("scan"));  // Laser scanner by default
+      const std::string source_type = get_parameter(source_name + ".type").as_string();
+
+      if (source_type == "scan") {
+        std::shared_ptr<Scan> s = std::make_shared<Scan>(
+          node, tf_buffer_, source_name, base_frame_id_,
+          transform_tolerance, max_time_shift_);
+
+        if (!s->init()) {
+          return false;
+        }
+
+        sources_.push_back(s);
+      } else {  // source_type == "pointcloud"
+        std::shared_ptr<PointCloud> p = std::make_shared<PointCloud>(
+          node, tf_buffer_, source_name, base_frame_id_,
+          transform_tolerance, max_time_shift_);
+
+        if (!p->init()) {
+          return false;
+        }
+
+        sources_.push_back(p);
+      }
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(get_logger(), "Error while getting parameters: %s", ex.what());
+    return false;
+  }
+
+  // Store polygons coordinates for a future re-use when publishing
+  std::vector<Point> poly;
+  for (std::shared_ptr<PolygonBase> polygon : polygons_) {
+    polygon->getPolygon(poly);
+    for (Point p : poly) {
+      geometry_msgs::msg::Point32 p_s;
+      p_s.x = p.x;
+      p_s.y = p.y;
+      p_s.z = 0.0;
+      polygons_pub_.points.push_back(p_s);
+    }
+  }
+
+  return true;
+}
+
+void CollisionMonitorNode::workerMain()
```

Comment:
This isn't a separate thread (right now) so `worker` isn't really an accurate name.

Reply:
Changed to `processMain()`

Reply:
Why not `process(Twist)`? That would also remove the need for `cmd_vel_in_` as a class member. Its only used in the `process` main function and can just be given it from the subscriber callback directly

Reply:
[Done](https://github.com/ros-planning/navigation2/pull/2982/files#diff-e1c76d88179c14a1cf399e94bebd4f5d17dedca39e09b3bbdc810aa11a705543R187-R190)